pydantic/_internal/_generate_schema.py

```diff
@@ -1215,11 +1215,15 @@ def _common_field_schema(  # C901
     ) -> _CommonField:
         # Update FieldInfo annotation if appropriate:
         FieldInfo = import_cached_field_info()
-        if has_instance_in_type(field_info.annotation, (ForwardRef, str)):
-            # TODO Can we use field_info.apply_typevars_map here? Shouldn't we use lenient=False?
-            evaluated = _typing_extra.eval_type(field_info.annotation, *self._types_namespace, lenient=True)
+        if not field_info.evaluated:
+            # TODO Can we use field_info.apply_typevars_map here?
+            try:
+                evaluated = _typing_extra.eval_type(field_info.annotation, *self._types_namespace)
+            except NameError as e:
+                raise PydanticUndefinedAnnotation.from_name_error(e) from e
```

Comment:
This pattern of reraising is quite common and could be made in a single function, but postponing as we might do more refactors on the class soon.

Reply:
Nice, looks cleaner.